src/mongo_odm/model.hpp

```diff
@@ -0,0 +1,148 @@
+   public:
+    /**
+     * Returns a copy of the underlying collection.
+     * @return The mongocxx::collection that this object wraps
+     */
+    static mongocxx::collection collection() {
+        return _coll.collection();
+    }
+
+    /**
+     * Sets the mongocxx::collection mapped to this serializable class.
+     *
+     * @param coll The mongocxx::collection object to be mapped to this class.
+     *
+     * @warning This must be called with a new mongocxx::collection instance for every thread for
+     *          the ODM to be thread-safe.
+     */
+    static void setCollection(const mongocxx::collection& coll) {
+        _coll = odm_collection<T>(coll);
+    }
+    static void setCollection(mongocxx::collection&& coll) {
+        _coll = odm_collection<T>(std::move(coll));
+    }
+
+    /**
+     * Performs an update in the database that saves the current T object instance to the
+     * collection mapped to this class.
+     *
+     * In the terms of the CRUD specification, this uses replaceOne with the _id as the sole
+     * argument to the query filter and upsert=true so that objects that aren't already in the
+     * collection are automatically inserted.
+     *
+     * @warning This will fully replace any object in the database with an _id equal to
+     *          this instance's _id.
+     */
+    void save() {
+        auto id_match_filter = bsoncxx::builder::stream::document{}
+                               << "_id" << _id << bsoncxx::builder::stream::finalize;
+        mongocxx::options::update opts{};
+        opts.upsert(true);
+
+        _coll.replace_one(id_match_filter.view(), *static_cast<T*>(this), opts);
+    };
+
+    /**
+     * Deletes this object from the mapped collection.
+     *
+     * In the terms of the CRUD specification, this uses deleteOne with the _id as the sole
+     * argument to the query filter.
+     */
+    void remove() {
+        auto id_match_filter = bsoncxx::builder::stream::document{}
+                               << "_id" << _id << bsoncxx::builder::stream::finalize;
+
+        _coll.collection().delete_one(id_match_filter.view());
+    }
+
+    ///
+    /// Finds the documents in this collection which match the provided filter.
+    ///
+    /// @param filter
+    ///   Document view representing a document that should match the query.
+    /// @param options
+    ///   Optional arguments, see mongocxx::options::find
+    ///
+    /// @return Cursor with deserialized objects from the collection.
+    /// @throws
+    ///   If the find failed, the returned cursor will throw mongocxx::exception::query when it
+    ///   is iterated.
+    ///
+    /// @see http://docs.mongodb.org/manual/core/read-operations-introduction/
+    ///
+    static deserializing_cursor<T> find(
+        bsoncxx::document::view_or_value&& filter,
+        const mongocxx::options::find& options = mongocxx::options::find()) {
+        return _coll.find(std::forward<bsoncxx::document::view_or_value>(filter),
```

Comment:
This should probably be
`return _coll.find(std::move<bsoncxx::document::view_or_value>(filter), options);`

`std::forward` is useful for universal references, but not here I think.